packages/sandpack-core/src/manager.ts

```diff
@@ -706,134 +712,224 @@ export default class Manager implements IEvaluator {
+        if (connectedPath.indexOf('/node_modules') !== 0) {
+          connectedPath = /^(\w|@\w|@-)/.test(shimmedPath)
+            ? pathUtils.join('/node_modules', shimmedPath)
+            : pathUtils.join(pathUtils.dirname(parentPath), shimmedPath);
+        }
 
-            return;
-          }
+        const isDependency = connectedPath.includes('/node_modules/');
 
-          this.cachedPaths[dirredPath][path] = foundPath;
+        connectedPath = connectedPath.replace('/node_modules/', '');
 
-          if (foundPath === '//empty.js') {
-            promiseResolve(getShimmedModuleFromPath(currentPath, path));
-            return;
-          }
+        if (!isDependency) {
+          throw new ModuleNotFoundError(shimmedPath, false, parentPath);
+        }
 
-          if (!this.transpiledModules[foundPath]) {
-            this.readFileSync(foundPath, (error, code) => {
-              if (error) {
-                promiseReject(error);
-                return;
-              }
+        const dependencyName = getDependencyName(connectedPath);
 
-              this.addModule({ path: foundPath, code: code || '' });
-              promiseResolve(this.transpiledModules[foundPath].module);
-            });
-          } else {
-            promiseResolve(this.transpiledModules[foundPath].module);
-          }
+        // TODO: fix the stack hack
+        if (
+          dependencyName &&
+          (this.manifest.dependencies.find(d => d.name === dependencyName) ||
+            this.manifest.dependencyDependencies[dependencyName] ||
+            this.manifest.contents[
+              `/node_modules/${dependencyName}/package.json`
+            ])
+        ) {
+          throw new ModuleNotFoundError(connectedPath, true, parentPath);
+        } else {
+          throw new DependencyNotFoundError(connectedPath, parentPath);
         }
-      );
-    });
+      }
+    }
+
+    if (resolvedPath === '//empty.js') {
+      return getShimmedModuleFromPath(parentPath, path);
+    }
+
+    return this.transpiledModules[resolvedPath].module;
   }
 
   // ALWAYS KEEP THIS METHOD IN SYNC WITH ASYNC VERSION
-  resolveModule(
-    path: string,
-    currentPath: string,
-    defaultExtensions: Array<string> = DEFAULT_EXTENSIONS
-  ): Module {
-    const dirredPath = pathUtils.dirname(currentPath);
+  resolveModule({
+    path,
+    parentPath = '/',
+    query = '',
+    defaultExtensions = DEFAULT_EXTENSIONS,
+  }: {
+    path: string;
+    parentPath?: string;
+    query?: string;
+    defaultExtensions?: Array<string>;
+  }): Module {
+    const esmoduleUrl = getESModuleUrl(parentPath, path);
+    // Handle ESModule import
+    if (esmoduleUrl) {
+      if (this.transpiledModules[esmoduleUrl]) {
+        return this.transpiledModules[esmoduleUrl].module;
+      }
+
+      throw new Error('Cannot download ESModule dependencies synchronously');
+    }
+
+    // Node modules?
+    if (isUrl(parentPath) && !esmoduleUrl) {
```

Comment:
Interesting, do you see this case happen as well?

Reply:
Oh yes it happens all the time as esmodules can be relative and absolute urls, this is also what we do in the codesandbox esmodule cdn

Reply:
Ahh, yes, I see. Would be good to put that scenario as a comment here for future readers (something simple like `esmodule file that requests a node module`).